modules/porous_flow/src/materials/PorousFlowFluidStateFlashBase.C

```diff
@@ -147,18 +178,156 @@ PorousFlowFluidStateFlashBase::computeQpProperties()
   setMaterialVectorSize();
 
   // Calculate all required thermophysical properties
-  thermophysicalProperties();
+  thermophysicalProperties(_fsp);
+  _saturation[_qp] = _fsp.saturation;
+  _porepressure[_qp] = _fsp.pressure;
+  _fluid_density[_qp] = _fsp.fluid_density;
+  _fluid_viscosity[_qp] = _fsp.fluid_viscosity;
+  _mass_frac[_qp] = _fsp.mass_fraction;
+
+  // Derivative of saturation wrt variables
+  _dsaturation_dvar[_qp][_gas_phase_number][_zvar[0]] = _fsp.dsaturation_dz[_gas_phase_number];
+  _dsaturation_dvar[_qp][_aqueous_phase_number][_zvar[0]] =
+      _fsp.dsaturation_dz[_aqueous_phase_number];
+  _dsaturation_dvar[_qp][_gas_phase_number][_pvar] = _fsp.dsaturation_dp[_gas_phase_number];
+  _dsaturation_dvar[_qp][_aqueous_phase_number][_pvar] = _fsp.dsaturation_dp[_aqueous_phase_number];
+
+  // Derivative of capillary pressure
+  Real seff = effectiveSaturation(_fsp.saturation[_aqueous_phase_number]);
+  Real dpc = dCapillaryPressure_dS(seff) * _dseff_ds;
+
+  // Derivative of porepressure wrt variables
+  if (_dictator.isPorousFlowVariable(_gas_porepressure_varnum))
+  {
+    for (unsigned int ph = 0; ph < _num_phases; ++ph)
+    {
+      _dporepressure_dvar[_qp][ph][_pvar] = 1.0;
+      if (!_nodal_material)
+        (*_dgradp_qp_dgradv)[_qp][ph][_pvar] = 1.0;
+    }
+
+    // The aqueous phase porepressure is also a function of liquid saturation,
+    // which depends on both gas porepressure and z
+    _dporepressure_dvar[_qp][_aqueous_phase_number][_pvar] +=
+        dpc * _dsaturation_dvar[_qp][_aqueous_phase_number][_pvar];
+    _dporepressure_dvar[_qp][_aqueous_phase_number][_zvar[0]] =
+        dpc * _dsaturation_dvar[_qp][_aqueous_phase_number][_zvar[0]];
+  }
+
+  // Calculate derivatives of material properties wrt primary variables
+  // Derivative of z wrt variables
+  std::vector<Real> dz_dvar;
+  dz_dvar.assign(_num_pf_vars, 0.0);
+  if (_dictator.isPorousFlowVariable(_z_varnum[0]))
+    dz_dvar[_zvar[0]] = 1.0;
+
+  // Derivatives of properties wrt primary variables
+  for (unsigned int v = 0; v < _num_pf_vars; ++v)
+  {
+    // Derivative of density in each phase
+    _dfluid_density_dvar[_qp][_aqueous_phase_number][v] =
+        _fsp.dfluid_density_dp[_aqueous_phase_number] *
+        _dporepressure_dvar[_qp][_aqueous_phase_number][v];
+    _dfluid_density_dvar[_qp][_aqueous_phase_number][v] +=
+        _fsp.dfluid_density_dT[_aqueous_phase_number] * _dtemperature_dvar[_qp][v];
+    _dfluid_density_dvar[_qp][_aqueous_phase_number][v] +=
+        _fsp.dfluid_density_dz[_aqueous_phase_number] * dz_dvar[v];
```

Comment:
Couldn't you also do a ``for (auto phase : phases)`` loop to remove more redundancy here and below?

And for even more deduplication (which may not be worth it) you could write a locally-scoped macro to do the dirty work for each variable (density, viscosity, etc.).